### ISSUE TITLE
boost::asio::bind_executor(...) result stored in the std::function problem

### ISSUE THREAD

**Issue by igsol** [chriskohlhoff/asio#278]
I'd appreciate if author could put attention on the issue https://github.com/boostorg/asio/issues/79
Thanks for help.

**Comment by arun11299** [comment]
Answering here as well.

The behaviour you are seeing is expected. Let me try to explain it in steps:

- Your `event_handler ` (std::function<...>) does not have an associated executor. So, the default executor is selected, which is the `system_executor`. For more details check `work_dispatcher` class.

- Now the executor being the `system_executor`, the work (`event_handler`) is dispatched to the executor which just calls the `event_handler` object.

- A call to `event_handler` is nothing but executing `executor_binder` functor which is the return value of `bind_executor`.

- The `executor_binder` is not responsible for running your wrapped task in the correct executor, it will just call the wrapped function in the context of the executor which called it i.e. the `system_executor`.

- If you ask, why #2 test (the now deprecated) works is because of what `strand::wrap` returns. Instead of just executing the wrapped function inside the context of the executor, it will again post it to the executor wrapped by the strand.

Hope this is clear.

**Comment by igsol** [comment]
Thanks for answer @arun11299.
I will put a look on the work_dispatcher class.

Actually what strand::wrap does is exactly what I need. But now handy method is deprecated for some unknown reason and suggested by documentation the boost::asio::bind_executor free function doesn't cover all cases when strand::wrap is useful. One of them I demonstrated in the example. Such deprecation could break our application and demand some more complex workaround instead of just trivial strand::wrap(...) call.

I'd like to know, could strand::wrap be un-depricated in future asio versions? :)
Or at least could documentation be more elaborated and suggest how to use strand, boost::asio::bind_executor with std::function?

**Comment by arun11299** [comment]
@igsol 
Why do you have to use `bind_executor`. Wouldnt following test example enough ?

`vfunc event_handler{[=] () { strand.post(lambda)}};`

**Comment by igsol** [comment]
The answer is here: http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/io_context__strand/wrap.html in the first sentence.

Example as a workaround is clear, but verbose in comparison with deprecated strand::wrap usage.

**Comment by igsol** [comment]
@arun11299 

> vfunc event_handler{[=] () { strand.post(lambda)}};

BTW http://www.boost.org/doc/libs/1_66_0/doc/html/boost_asio/reference/io_context__strand/post/overload2.html states: "Deprecated: Use post."

**Comment by arthurprs** [comment]
I wish I found this earlier. It took me a longggg time to track down a memory corruption bug which was caused by my misinterpretation of `bind_executor`. The documentation could use some clarification instead of pointing straight to bind_executor.

IMHO strand.wrap is what most people want.

**Comment by arthurprs** [comment]
The alternative (as proposed) gets exponentially cumbersome as the handler of arguments in the handler increase.

**Comment by klmurphy72** [comment]
@arthurprs  said:
> The documentation could use some clarification instead of pointing straight to bind_executor.
>
>IMHO strand.wrap is what most people want.


I very much agree with this.  They are not equivalent, so having the docs point you straight from one to the other without explaining that the two do not behave identically is misleading.

Thanks to this conversation it only took me a longgg time to fix my problem as apposed to @arthurprs longggg time. :)

**Comment by dzhioev** [comment]
> I wish I found this earlier. It took me a longggg time to track down a memory corruption bug which was caused by my misinterpretation of `bind_executor`. The documentation could use some clarification instead of pointing straight to bind_executor.

Totally agree. Documentation is super misleading now.

**Comment by arun11299** [comment]
Agreed. The framework is becoming super complex and probably not a good choice anymore if you are managing a specialized server to be written using it.

**Comment by raJeev-M** [comment]
It's been a long time since I looked into Asio, But to recall from my memory, I have heard from the author that...

1) Asio is a tool-kit not a  framework.

2) I feel, the name given to the Identifier 'bind_executor' is pretty self-explanatory.

An Executor, simply contains rules-- how you want your completion-handler to be executed-- now, think of "Bind..Executor" along with your completion-handler.

Each Executor is associated with an Execution-context-- context of your completion handler's execution-- EC maintains a queue of completion-handler.

I would say that, the code is well documented.

And header-only Asio is a gift-- don't pay for things you don't use.

**Comment by dzhioev** [comment]
@arun11299 I'm trying to implement now deprecated strand::wrap without using deprecated parts. Here is what I did:
`template <typename F>                                                                                   
decltype(auto) wrap(boost::asio::strand<boost::asio::io_context>& strand, F&& callback) {               
  return boost::asio::bind_executor(                                                                    
      strand,                                                                                           
      [strand_ptr=&strand, callback = std::move(callback)](auto&&... args) mutable {                    
        boost::asio::dispatch(                                                                          
            *strand_ptr,                                                                                
            [callback = std::move(callback),                                                            
             args_tuple = std::make_tuple(std::forward<decltype(args)>(args)...)]() mutable {           
              return std::apply(std::move(callback), std::move(args_tuple));                            
            });                                                                                         
      });                                                                                               
} `
Could you please confirm that this is correct?

**Comment by DonOregano** [comment]
Just here to say that I was bitten by this "documentation flaw". I was attempting to remove usage of deprecated features from a fairly large asio-using codebase, and I followed the Networking TS compatibility "guide" fairly blindly. I agree completely with above commentors that the reference to bind_executor from wrap is very misleading. I would instead say something like "there is no equivalent to strand::wrap, but it can be recreated with <insert above  code by @dzhioev>.

**Comment by DonOregano** [comment]
Could we have some closure on this issue? The situation is obviously confusing to quite a few people, and for example the code in https://www.boost.org/doc/libs/1_79_0/doc/html/boost_asio/tutorial/tuttimer5.html does not help clarify what is going on. I am not really sure I have understood the issue at all. Which is preventing me from removing deprecated asio calls from my project.

Have I understood correctly if I say that the difference between strand.wrap and bind_executor is that the result of the latter cannot be stored in a std::function?

**Comment by klemens-morgenstern** [comment]
Asio has associators, which are type-trait based. If you call `get_associated_executor(X)` it will use a type_trait to determine if there's an associated executor. Internally the async operation will use that for completion.

`std::function` doesn't have one, but whatever `bind_associate_executor` has, does. If you want a type-erased completion handler, maybe `any_completion_handler` does the trick for you.

I think the `strand.wrap` function predates the associators.
